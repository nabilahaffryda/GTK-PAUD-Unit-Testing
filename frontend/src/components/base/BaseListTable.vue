<template>
  <v-data-table
    :hide-default-footer="!usePaging"
    :items="item || []"
    :options.sync="options"
    :items-per-page="limit"
    :server-items-length="total"
    :loading="loading"
    disable-sort
    :hide-default-header="hideHeader"
    :headers="headers"
    :footer-props="{ itemsPerPageOptions: [limit] }"
  >
    <template v-slot:header="{ props: { headers } }">
      <thead>
        <slot name="header" :headers="headers"></slot>
      </thead>
    </template>
    <template v-slot:body="{ items }">
      <tbody>
        <template v-if="items.length">
          <tr v-for="(item, i) in items" :key="i" class="transparent">
            <slot :item="item" :index="i" />
          </tr>
        </template>
        <template v-else>
          <tr>
            <td :colspan="headers.length">
              {{
                keyword
                  ? `Data Pencarian ${title}${keyword ? ` dengan kata kunci ${keyword}` : ''} tidak ditemukan`
                  : 'Belum ada data yang ditemukan'
              }}
            </td>
          </tr>
        </template>
      </tbody>
    </template>
  </v-data-table>
</template>

<script>
export default {
  name: 'BaseListTable',
  props: {
    usePaging: {
      type: Boolean,
      default: true,
    },
    loading: {
      type: Boolean,
      default: true,
    },
    hideHeader: {
      type: Boolean,
      default: false,
    },
    headers: {
      type: Array,
      default: () => [],
    },
    item: {
      type: Array,
      default: () => [],
    },
    opt: {
      type: Object,
      default: () => {},
    },
    total: {
      type: Number,
      default: 0,
    },
    limit: {
      type: Number,
      default: 10,
    },
    title: {
      type: String,
      default: '',
    },
    keyword: {
      type: String,
      default: '',
    },
  },
  data() {
    return {
      options: this.opt || {},
    };
  },
  methods: {
    fetch() {
      this.$emit('fetch', {
        page: this.options.page,
        limit: this.options.itemsPerPage,
      });
    },
  },
  watch: {
    options: {
      handler() {
        this.fetch();
      },
      deep: true,
    },
  },
};
</script>

<style scoped></style>
