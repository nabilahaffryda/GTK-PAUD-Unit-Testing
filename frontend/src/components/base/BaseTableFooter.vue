<template>
  <div class="text-center" style="width: 100%">
    <template v-if="useNumber">
      <v-pagination
        :circle="circle"
        color="secondary"
        v-model="page"
        :length="pageTotal"
        prev-icon="mdi-chevron-left"
        next-icon="mdi-chevron-right"
        :total-visible="5"
      >
      </v-pagination>
    </template>
    <template v-else>
      <v-btn
        fab
        color="secondary"
        class="ma-1"
        x-small
        :disabled="Number(page) === 1"
        :depressed="Number(page) === 1"
        @click="prev"
      >
        <v-icon>mdi-chevron-left</v-icon>
      </v-btn>
      <v-btn
        fab
        color="secondary"
        class="ma-1"
        x-small
        :disabled="Number(total) < 10"
        :depressed="Number(total) < 10"
        @click="next"
      >
        <v-icon>mdi-chevron-right</v-icon>
      </v-btn>
    </template>
  </div>
</template>

<script>
export default {
  name: 'BaseTableFooter',
  props: {
    pageTotal: {
      type: Number,
      default: 0,
    },
    total: {
      type: Number,
      default: 0,
    },
    circle: {
      type: Boolean,
      default: true,
    },
    useNumber: {
      type: Boolean,
      default: true,
    },
  },
  data() {
    return {
      page: 1,
    };
  },
  methods: {
    next() {
      this.page = this.page + 1;
    },
    prev() {
      this.page = this.page > 1 ? this.page - 1 : this.page;
    },
    changePage() {
      this.$emit('changePage', this.page);
    },
  },
  watch: {
    page: 'changePage',
  },
};
</script>

<style scoped></style>
