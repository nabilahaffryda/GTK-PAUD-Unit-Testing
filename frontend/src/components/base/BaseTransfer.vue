<template>
  <v-row>
    <v-col>
      <h1 class="title mb-2 secondary--text">Data Pilihan {{ title }}</h1>
      <v-card flat outlined>
        <v-card-title class="pb-0 mb-0">
          <slot name="filter" />
          <v-toolbar flat>
            <v-toolbar-title>Pilih {{ title }}</v-toolbar-title>
            <v-spacer></v-spacer>
            <v-btn small icon @click="fetchData">
              <v-icon color="secondary">mdi-reload</v-icon>
            </v-btn>
            <v-btn small icon @click="pickAll">
              <v-icon color="secondary">mdi-arrow-right-bold</v-icon>
            </v-btn>
          </v-toolbar>
        </v-card-title>
        <v-divider></v-divider>
        <v-card-text>
          <base-list-table
            :hideHeader="true"
            :loading="dataLoading"
            :item="dataList"
            :total="totalData"
            :usePaging="false"
            @fetch="fetchData"
          >
            <template slot-scope="{ item }">
              <td
                :class="[
                  'lighten-3',
                  {
                    grey: targetList.findIndex((target) => target[keyAttr] === item[keyAttr]) > -1,
                  },
                ]"
              >
                <v-list-item dense class="px-0">
                  <v-list-item-content>
                    <slot :item="item" />
                  </v-list-item-content>
                  <v-list-item-action-text>
                    <template v-if="targetList.findIndex((target) => target[keyAttr] === item[keyAttr]) > -1">
                      <v-icon color="secondary">mdi-check</v-icon>
                    </template>
                    <template v-else>
                      <v-btn icon @click="onSelect(item)">
                        <v-icon color="secondary">mdi-arrow-right-bold</v-icon>
                      </v-btn>
                    </template>
                  </v-list-item-action-text>
                </v-list-item>
              </td>
            </template>
          </base-list-table>
        </v-card-text>
        <v-card-actions>
          <base-table-footer :pageTotal="pageTotalData" @changePage="onChangePageData"></base-table-footer>
        </v-card-actions>
      </v-card>
    </v-col>
    <v-col>
      <h1 class="title mb-2 secondary--text"> Data {{ title }} untuk Paket Baru </h1>
      <v-card flat outlined>
        <v-card-title class="pb-0 mb-0">
          <v-toolbar flat>
            <v-toolbar-title>{{ totalTarget }} {{ title }} Terpilih</v-toolbar-title>
            <v-spacer></v-spacer>
            <v-btn small icon @click="deleteAll">
              <v-icon color="secondary">mdi-close</v-icon>
            </v-btn>
          </v-toolbar>
        </v-card-title>
        <v-divider></v-divider>
        <v-card-text>
          <base-list-table
            :hideHeader="true"
            :loading="false"
            :item="targetList"
            :total="totalTarget"
            :usePaging="false"
          >
            <template slot-scope="{ item }">
              <td>
                <v-list-item dense class="px-0">
                  <v-list-item-content>
                    <slot :item="item" />
                  </v-list-item-content>
                  <v-list-item-action-text>
                    <v-btn icon @click="onDelete(item)">
                      <v-icon color="secondary">mdi-close</v-icon>
                    </v-btn>
                  </v-list-item-action-text>
                </v-list-item>
              </td>
            </template>
          </base-list-table>
        </v-card-text>
        <v-card-actions>
          <base-table-footer :pageTotal="pageTotalTarget" @changePage="onChangePageTarget"></base-table-footer>
        </v-card-actions>
      </v-card>
    </v-col>
  </v-row>
</template>

<script>
export default {
  name: 'BaseTransfer',
  props: {
    title: {
      type: String,
      required: true,
    },
    keyAttr: {
      type: String,
      required: true,
    },
    dataLoading: {
      type: Boolean,
      default: true,
    },
    targetLoading: {
      type: Boolean,
      default: true,
    },
    dataList: {
      type: Array,
      default: () => [],
    },
    targetList: {
      type: Array,
      default: () => [],
    },
    totalData: {
      type: Number,
      default: 10,
    },
    totalTarget: {
      type: Number,
      default: 10,
    },
    pageTotalData: {
      type: Number,
      default: 10,
    },
    pageTotalTarget: {
      type: Number,
      default: 10,
    },
  },
  methods: {
    fetchData() {
      this.$emit('fetch');
    },
    pickAll() {
      this.$emit('pickall');
    },
    deleteAll() {
      this.$emit('deleteall');
    },
    onChangePageData(page) {
      this.$emit('changePageData', page);
    },
    onChangePageTarget(page) {
      this.$emit('changePageTarget', page);
    },
    onSelect(data) {
      this.$emit('selected', data);
    },
    onDelete(data) {
      this.$emit('delete', data);
    },
  },
};
</script>

<style scoped></style>
