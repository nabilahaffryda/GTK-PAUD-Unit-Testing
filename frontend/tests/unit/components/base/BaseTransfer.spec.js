import { mount, createLocalVue } from "@vue/test-utils";
import BaseTransfer from "@/components/base/BaseTransfer.vue";
import Vuetify from 'vuetify';
import Vue from 'vue';
import BaseListTable from "@/components/base/BaseListTable.vue";
import BaseTableFooter from "@/components/base/BaseTableFooter.vue";

Vue.use(Vuetify)
const localVue = createLocalVue();

describe('BaseTransfer.vue', () => {
    let vuetify
    beforeEach(() => {
        vuetify = new Vuetify()
    })
    test('render transfer', () => {
        const VCard = {
            props: [
                'dataLoading',
                'targetLoading',
                'dataList',
                'targetList',
                'totalData',
                'totalTarget',
                'pageTotalData',
                'pageTotalTarget'],
            template: '<div><slot :item="dataList" /></div>'
        }

        const wrapper = mount(BaseTransfer, {
            localVue,
            vuetify,
            stubs: {
                VCard, BaseListTable, BaseTableFooter
            },
            propsData: {
                title: 'transfer',
                keyAttr: 'key1',
            }
        });
        expect(wrapper).toMatchSnapshot();
    })
})
