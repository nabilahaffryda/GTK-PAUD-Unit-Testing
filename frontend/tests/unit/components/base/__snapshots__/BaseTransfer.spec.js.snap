// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BaseTransfer.vue render transfer 1`] = `
<div class="row">
  <div class="col">
    <h1 class="title mb-2 secondary--text">Data Pilihan transfer</h1>
    <div flat="" outlined="">
      <div class="v-card__title pb-0 mb-0">
        <header class="v-sheet theme--light v-toolbar v-toolbar--flat" style="height: 64px;">
          <div class="v-toolbar__content" style="height: 64px;">
            <div class="v-toolbar__title">Pilih transfer</div>
            <div class="spacer"></div> <button type="button" class="v-btn v-btn--icon v-btn--round theme--light v-size--small"><span class="v-btn__content"><i aria-hidden="true" class="v-icon notranslate mdi mdi-reload theme--light secondary--text"></i></span></button> <button type="button" class="v-btn v-btn--icon v-btn--round theme--light v-size--small"><span class="v-btn__content"><i aria-hidden="true" class="v-icon notranslate mdi mdi-arrow-right-bold theme--light secondary--text"></i></span></button>
          </div>
        </header>
      </div>
      <hr role="separator" aria-orientation="horizontal" class="v-divider theme--light">
      <div class="v-card__text">
        <div class="v-data-table theme--light">
          <div class="v-data-table__wrapper">
            <table>
              <colgroup></colgroup>
              <thead></thead>
              <thead>
                <tr class="v-data-table__progress">
                  <th colspan="0" class="column">
                    <div role="progressbar" aria-valuemin="0" aria-valuemax="100" class="v-progress-linear v-progress-linear--absolute v-progress-linear--visible theme--light" style="height: 4px;">
                      <div class="v-progress-linear__background primary" style="opacity: 0.3; left: 0%; width: 100%;"></div>
                      <div class="v-progress-linear__buffer"></div>
                      <transition-stub name="fade-transition">
                        <div class="v-progress-linear__indeterminate v-progress-linear__indeterminate--active">
                          <div class="v-progress-linear__indeterminate long primary"></div>
                          <div class="v-progress-linear__indeterminate short primary"></div>
                        </div>
                      </transition-stub>
                    </div>
                  </th>
                </tr>
              </thead>
              <tbody>
                <tr>
                  <td colspan="0">
                    Belum ada data yang ditemukan
                  </td>
                </tr>
              </tbody>
            </table>
          </div>
        </div>
      </div>
      <div class="v-card__actions">
        <div class="text-center" style="width: 100%;">
          <nav role="navigation" aria-label="Pagination Navigation">
            <ul class="v-pagination v-pagination--circle theme--light">
              <li><button disabled="disabled" type="button" aria-label="Previous page" class="v-pagination__navigation v-pagination__navigation--disabled"><i aria-hidden="true" class="v-icon notranslate mdi mdi-chevron-left theme--light"></i></button></li>
              <li><button type="button" aria-current="true" aria-label="Current Page, Page 1" class="v-pagination__item v-pagination__item--active secondary">1</button></li>
              <li><button type="button" aria-label="Goto Page 2" class="v-pagination__item">2</button></li>
              <li><span class="v-pagination__more">...</span></li>
              <li><button type="button" aria-label="Goto Page 9" class="v-pagination__item">9</button></li>
              <li><button type="button" aria-label="Goto Page 10" class="v-pagination__item">10</button></li>
              <li><button type="button" aria-label="Next page" class="v-pagination__navigation"><i aria-hidden="true" class="v-icon notranslate mdi mdi-chevron-right theme--light"></i></button></li>
            </ul>
          </nav>
        </div>
      </div>
    </div>
  </div>
  <div class="col">
    <h1 class="title mb-2 secondary--text"> Data transfer untuk Paket Baru </h1>
    <div flat="" outlined="">
      <div class="v-card__title pb-0 mb-0">
        <header class="v-sheet theme--light v-toolbar v-toolbar--flat" style="height: 64px;">
          <div class="v-toolbar__content" style="height: 64px;">
            <div class="v-toolbar__title">10 transfer Terpilih</div>
            <div class="spacer"></div> <button type="button" class="v-btn v-btn--icon v-btn--round theme--light v-size--small"><span class="v-btn__content"><i aria-hidden="true" class="v-icon notranslate mdi mdi-close theme--light secondary--text"></i></span></button>
          </div>
        </header>
      </div>
      <hr role="separator" aria-orientation="horizontal" class="v-divider theme--light">
      <div class="v-card__text">
        <div class="v-data-table theme--light">
          <div class="v-data-table__wrapper">
            <table>
              <colgroup></colgroup>
              <thead></thead>
              <tbody>
                <tr>
                  <td colspan="0">
                    Belum ada data yang ditemukan
                  </td>
                </tr>
              </tbody>
            </table>
          </div>
        </div>
      </div>
      <div class="v-card__actions">
        <div class="text-center" style="width: 100%;">
          <nav role="navigation" aria-label="Pagination Navigation">
            <ul class="v-pagination v-pagination--circle theme--light">
              <li><button disabled="disabled" type="button" aria-label="Previous page" class="v-pagination__navigation v-pagination__navigation--disabled"><i aria-hidden="true" class="v-icon notranslate mdi mdi-chevron-left theme--light"></i></button></li>
              <li><button type="button" aria-current="true" aria-label="Current Page, Page 1" class="v-pagination__item v-pagination__item--active secondary">1</button></li>
              <li><button type="button" aria-label="Goto Page 2" class="v-pagination__item">2</button></li>
              <li><span class="v-pagination__more">...</span></li>
              <li><button type="button" aria-label="Goto Page 9" class="v-pagination__item">9</button></li>
              <li><button type="button" aria-label="Goto Page 10" class="v-pagination__item">10</button></li>
              <li><button type="button" aria-label="Next page" class="v-pagination__navigation"><i aria-hidden="true" class="v-icon notranslate mdi mdi-chevron-right theme--light"></i></button></li>
            </ul>
          </nav>
        </div>
      </div>
    </div>
  </div>
</div>
`;
